File: app/static/filter-composer.js
/* Compositeur d'expressions de filtre Akvorado — « à la Palo Alto ».
 *
 * LE GESTE QUE CE FICHIER RENVERSE : jusqu'ici, enregistrer un filtre demandait
 * de connaître par cœur la syntaxe d'Akvorado ET le nom exact de ses 61
 * colonnes, puis de les taper dans un champ texte nu. Ici on part de ce qui
 * EXISTE dans les derniers flux : on cherche, on clique un champ, on clique une
 * valeur réellement observée, et l'expression se construit.
 *
 * POURQUOI CE FICHIER EXISTE plutôt qu'un `<script>` ou des `onclick=` dans le
 * template : la CSP du service est `script-src 'self'`, SANS `unsafe-inline` ni
 * `unsafe-eval`. Tout JS inline — bloc `<script>` comme attribut `onclick` — est
 * BLOQUÉ par le navigateur, et il l'est en silence du point de vue de
 * l'application : la page s'affiche, les boutons sont là, rien ne réagit au
 * clic. Défaut mesuré deux fois sur ce projet (2026-08-06 : `onchange=` du
 * formulaire d'interface, `onchange=` du sélecteur de fenêtre).
 *
 * POURQUOI la délégation d'événements sur `document` plutôt qu'un écouteur par
 * bouton : les cartes de champs sont RÉINJECTÉES par htmx à chaque frappe dans
 * la barre de recherche. Des écouteurs posés au chargement seraient perdus sur
 * le premier fragment remplacé — le compositeur cesserait de réagir dès la
 * première recherche, c'est-à-dire immédiatement.
 *
 * CE QUE CE FICHIER NE FAIT PAS, DÉLIBÉRÉMENT :
 * - il ne décide JAMAIS si une valeur a besoin de guillemets. C'est la console
 *   Akvorado qui le dit, via `data-composer-quoted`. Une règle devinée ici
 *   (« un nombre n'a pas besoin de quotes ») divergerait de la version
 *   d'Akvorado déployée et produirait des filtres refusés.
 * - il ne valide JAMAIS la grammaire. Le verdict vient de
 *   `/config/filters/validate`, seule autorité sur sa propre syntaxe.
 * - il n'est JAMAIS un passage obligé : le champ d'expression reste un
 *   `<textarea>` ordinaire dans un `<form>` ordinaire. Script non exécuté, le
 *   formulaire d'ajout reste soumissible à la main (voir « REPLI SANS JS »).
 */
(function () {
  "use strict";

  /* Racine du compositeur. Tout est cherché SOUS elle : la page de section peut
   * contenir d'autres compositeurs (celui des étiquettes de `visualize_defaults`,
   * `field-composer.js`), et deux scripts qui se disputent les mêmes sélecteurs
   * produiraient des insertions dans le mauvais champ. */
  var ROOT_SELECTOR = "[data-filter-composer]";
  var EXPRESSION_SELECTOR = "[data-composer-expression]";

  function root() {
    return document.querySelector(ROOT_SELECTOR);
  }

  function expressionField() {
    /* Recherche dans TOUT le document, et non dans la racine du compositeur.
     *
     * DÉFAUT MESURÉ AU NAVIGATEUR (2026-08-06) : la zone d'expression vit dans
     * le FORMULAIRE D'AJOUT, en dehors du bloc de composition — c'est même sa
     * place naturelle, puisque c'est elle qui est postée. Chercher dans la
     * racine ne la trouvait donc jamais : chaque clic sur un champ, une valeur
     * ou un opérateur ne produisait RIEN, sans erreur en console ni à l'écran.
     * Le compositeur s'affichait parfaitement et ne composait pas.
     *
     * Sonde qui l'a montré : `racine.contains(textarea)` -> false.
     *
     * Le repli sur la racine est conservé pour le cas où plusieurs zones
     * coexisteraient un jour sur la même page. */
    var direct = document.querySelector(EXPRESSION_SELECTOR);
    if (direct) {
      return direct;
    }
    var host = root();
    return host ? host.querySelector(EXPRESSION_SELECTOR) : null;
  }

  /* ---------------------------------------------------------------------
   * Insertion dans l'expression
   * ------------------------------------------------------------------ */

  /* Un fragment s'insère À LA POSITION DU CURSEUR, pas en fin de champ.
   *
   * POURQUOI : l'expression se construit rarement de gauche à droite d'un seul
   * jet. On tape une première comparaison suivie de `AND `, on se rend compte
   * qu'il manque une parenthèse ouvrante au tout début, on clique `(`. Une
   * insertion systématique en fin de champ collerait la parenthèse APRÈS le
   * `AND` — donc au mauvais endroit — et l'utilisateur devrait défaire à la
   * main ce que le bouton vient de faire. Le compositeur doit obéir au curseur,
   * jamais le déplacer d'autorité. */
  function insert(fragment) {
    var field = expressionField();
    if (!field) {
      return;
    }

    var value = field.value;
    /* `selectionStart` vaut `null` sur certains types de champ et quand le
     * champ n'a jamais reçu le focus : on retombe alors sur la fin du texte,
     * qui est le comportement attendu pour une première insertion. */
    var start = typeof field.selectionStart === "number" ? field.selectionStart : value.length;
    var end = typeof field.selectionEnd === "number" ? field.selectionEnd : value.length;

    var avant = value.slice(0, start);
    var apres = value.slice(end);

    /* Espacement calculé, jamais concaténation brute : `SrcPort` collé à un
     * `AND` précédent donnerait `ANDSrcPort`, refusé par Akvorado pour une
     * raison que rien à l'écran n'expliquerait. On n'ajoute d'espace que s'il
     * en manque un — sans quoi cliquer trois champs de suite produirait des
     * doubles espaces qui, eux, sont inoffensifs mais donnent une expression
     * d'apparence négligée. */
    var separateurGauche = avant.length > 0 && !/\s$/.test(avant) ? " " : "";
    var separateurDroite = apres.length > 0 && !/^\s/.test(apres) ? " " : "";

    var insere = separateurGauche + fragment + separateurDroite;
    field.value = avant + insere + apres;

    /* Le curseur se repositionne APRÈS le fragment inséré (avant l'espace de
     * droite) : la frappe suivante continue naturellement l'expression au lieu
     * de repartir du début du champ. */
    var curseur = avant.length + separateurGauche.length + fragment.length;
    if (typeof field.setSelectionRange === "function") {
      field.setSelectionRange(curseur, curseur);
    }
    field.focus();

    /* L'événement `input` est ÉMIS À LA MAIN : une écriture par script sur
     * `field.value` n'en déclenche AUCUN. Sans cette ligne, la validation
     * htmx (câblée sur `input changed delay:500ms`) ne se déclencherait jamais
     * sur les fragments insérés au clic — le verdict resterait figé sur l'état
     * précédent, donc MENTIRAIT sur l'expression réellement affichée. C'est le
     * motif du « zéro silencieux » : un état affiché qui ne correspond plus à
     * la mesure. */
    field.dispatchEvent(new Event("input", { bubbles: true }));
  }

  /* Une valeur ne se cite QUE si la console l'a dit. `data-composer-quoted`
   * vaut "1" (guillemets requis, ex. `ExporterName = 'proxy-frontal'`) ou "0"
   * (valeur nue, ex. `SrcPort = 443`).
   *
   * ABSENCE DE L'ATTRIBUT : la route `/config/filters/fields` rend les valeurs
   * de l'échantillon SANS `data-composer-quoted` — seule
   * `/config/filters/values` le porte. On ne devine pas pour autant : on
   * retombe sur la citation, parce qu'Akvorado accepte `SrcPort = '443'` (une
   * chaîne se compare à un nombre) alors qu'il REFUSE
   * `ExporterName = proxy-frontal` (un identifiant nu n'est pas une chaîne). En cas
   * de doute, citer est le choix qui échoue le moins. */
  function quoteIfNeeded(valeur, quotedAttr) {
    if (quotedAttr === "0") {
      return valeur;
    }
    /* Une apostrophe dans la valeur casserait la chaîne et changerait le sens
     * de l'expression. On la double, comme le fait SQL. */
    return "'" + String(valeur).replace(/'/g, "''") + "'";
  }

  /* ---------------------------------------------------------------------
   * Anti double-insertion
   * ------------------------------------------------------------------ */

  /* Un double-clic sur un champ produit DEUX événements `click`. Sans garde,
   * l'expression recevrait `SrcAddr SrcAddr` — et l'utilisateur, qui a juste
   * cliqué un peu vite, verrait une expression fausse sans comprendre pourquoi.
   *
   * Le garde porte sur (élément + fragment) et non sur un simple verrou global
   * temporel : cliquer volontairement deux fois `AND` à 300 ms d'intervalle est
   * un geste LÉGITIME sur un bouton d'opérateur, alors que deux `click` sur le
   * MÊME bouton de champ en moins de 350 ms ne l'est jamais. */
  var DOUBLE_CLIC_MS = 350;
  var dernierClic = { cible: null, fragment: null, at: 0 };

  function estDoublon(cible, fragment) {
    var maintenant = Date.now();
    var doublon =
      dernierClic.cible === cible &&
      dernierClic.fragment === fragment &&
      maintenant - dernierClic.at < DOUBLE_CLIC_MS;
    dernierClic = { cible: cible, fragment: fragment, at: maintenant };
    return doublon;
  }

  /* ---------------------------------------------------------------------
   * Délégation des clics
   * ------------------------------------------------------------------ */

  document.addEventListener("click", function (event) {
    var target = event.target;
    if (!target || typeof target.closest !== "function") {
      return;
    }

    /* Le compositeur vit dans une page qui contient d'autres blocs cliquables.
     * On ne réagit qu'aux clics DANS le compositeur. */
    var host = target.closest(ROOT_SELECTOR);
    if (!host) {
      return;
    }

    /* --- Une valeur : insère `Champ = valeur` (ou `Champ = 'valeur'`) ---
     * Testé AVANT le champ : un bouton de valeur porte lui aussi
     * `data-composer-field` (il a besoin de savoir à quelle colonne il
     * appartient). Dans l'ordre inverse, cliquer une valeur insérerait le seul
     * nom de colonne — le clic le plus utile du compositeur ne ferait que la
     * moitié de son travail. */
    var valueBtn = target.closest("[data-composer-value]");
    if (valueBtn && host.contains(valueBtn)) {
      var champ = valueBtn.getAttribute("data-composer-field") || "";
      var valeur = valueBtn.getAttribute("data-composer-value") || "";
      var quoted = valueBtn.getAttribute("data-composer-quoted");
      var fragment = champ + " = " + quoteIfNeeded(valeur, quoted);
      if (!estDoublon(valueBtn, fragment)) {
        insert(fragment);
      }
      event.preventDefault();
      return;
    }

    /* --- Un champ : insère le seul nom de colonne ---
     * Sans opérateur ni valeur : l'utilisateur enchaîne avec le bouton `=` puis
     * une valeur, ou tape ce qu'il veut. Insérer `SrcAddr = ` d'autorité
     * imposerait l'égalité alors que `SrcAddr IN (...)` est tout aussi courant. */
    var fieldBtn = target.closest("[data-composer-field]");
    if (fieldBtn && host.contains(fieldBtn)) {
      var nom = fieldBtn.getAttribute("data-composer-field") || "";
      if (nom && !estDoublon(fieldBtn, nom)) {
        insert(nom);
      }
      event.preventDefault();
      return;
    }

    /* --- Un opérateur --- */
    var opBtn = target.closest("[data-composer-operator]");
    if (opBtn && host.contains(opBtn)) {
      var op = opBtn.getAttribute("data-composer-operator") || "";
      if (op && !estDoublon(opBtn, op)) {
        insert(op);
      }
      event.preventDefault();
      return;
    }

    /* --- Vider l'expression ---
     * POURQUOI un bouton dédié plutôt que « sélectionner tout puis supprimer » :
     * une expression composée au clic fait vite 120 caractères, et la reprendre
     * de zéro est un geste courant. Pas de confirmation : rien n'est enregistré
     * à ce stade, l'expression n'est qu'un brouillon. */
    var clearBtn = target.closest("[data-composer-clear]");
    if (clearBtn && host.contains(clearBtn)) {
      var field = expressionField();
      if (field) {
        field.value = "";
        field.focus();
        field.dispatchEvent(new Event("input", { bubbles: true }));
      }
      event.preventDefault();
      return;
    }
  });

  /* ---------------------------------------------------------------------
   * REPLI SANS JS — ce que le script MASQUE, jamais ce qu'il révèle
   * ------------------------------------------------------------------ */

  /* Le compositeur est une AIDE, pas un passage obligé. Le formulaire d'ajout
   * est un `<form>` ordinaire avec un `<textarea name="content">` : sans ce
   * script, on tape l'expression à la main et on l'enregistre — exactement
   * comme avant ce chantier, sans aucune perte de capacité.
   *
   * Cette marque `data-filter-composer-ready` est posée par le SCRIPT, à
   * l'exécution. Le CSS s'en sert pour masquer la mention « composition au clic
   * indisponible » et pour révéler les commandes qui n'ont de sens qu'avec du
   * JS (opérateurs, bouton vider).
   *
   * LE SENS DE LA MARQUE EST CRUCIAL : les commandes JS sont rendues par le
   * serveur puis MASQUÉES si le script ne tourne pas — jamais l'inverse. Posées
   * masquées et révélées par le script, elles resteraient invisibles le seul
   * jour où ça compte : celui où le script échoue. Même raisonnement que le
   * bouton de repli du sélecteur de fenêtre (`autosubmit-fallback`, mesuré
   * 2026-08-06). */
  function marquerPret() {
    var host = root();
    if (host) {
      host.setAttribute("data-filter-composer-ready", "1");
    }
  }

  /* Posée aux deux moments : au chargement initial, et après chaque swap htmx —
   * le panneau de section entier peut être réinjecté après un changement mis en
   * file, ce qui emporterait la marque avec l'ancien DOM. */
  document.addEventListener("DOMContentLoaded", marquerPret);
  document.addEventListener("htmx:afterSwap", marquerPret);
  /* `defer` garantit l'exécution après l'analyse du document, mais le script
   * peut aussi être évalué APRÈS `DOMContentLoaded` (cache, réinjection). Cet
   * appel direct couvre ce cas — sans lui, la marque ne serait jamais posée et
   * l'écran afficherait à tort « composition indisponible ». */
  marquerPret();
})();

File: app/routers/filter_composer.py
"""Compositeur de filtres Akvorado — à la manière d'un Palo Alto.

Jusqu'ici, enregistrer un filtre demandait de connaître par cœur la syntaxe
d'Akvorado ET le nom exact de ses 61 colonnes, puis de taper l'expression dans
un champ texte nu. Aucune aide, aucune vérification avant enregistrement : une
faute de frappe ne se découvrait qu'en ouvrant la console.

Cet écran renverse le geste. On part de ce qui EXISTE — un échantillon des
derniers flux — on clique sur un champ, on choisit une valeur réellement
observée, et l'expression se construit toute seule. La validation se fait au
fil de la frappe, par Akvorado lui-même.

TROIS SOURCES, chacune la seule autorité sur son sujet :

- `app.services.flow_sample` : l'échantillon lu dans ClickHouse. C'est lui qui
  dit quels champs portent des valeurs, et lesquelles.
- `app.clients.akvorado_console` : l'autocomplétion et la validation. La
  console interroge ClickHouse pour proposer les valeurs réelles, et c'est elle
  qui tranche sur sa propre grammaire — jamais une regex maison, qui
  divergerait de la version d'Akvorado déployée.
- `app.services.config_sections` : le YAML des filtres, seule source de vérité
  pour ce qui est enregistré (la console les resynchronise destructivement à
  son démarrage).

DÉGRADATION VISIBLE : ClickHouse ou la console peuvent ne pas répondre. Chaque
route rend alors un état DISTINCT — « je n'ai pas pu interroger » — et jamais
une liste vide, qui serait indistinguable de « aucun champ ne correspond ».
"""

from __future__ import annotations

import logging
from html import escape
from typing import Annotated, Any

from fastapi import APIRouter, Depends, Form
from fastapi.responses import HTMLResponse

from app.clients.akvorado_console import ConsoleUnavailable, complete_filter, validate_filter
from app.services.flow_sample import (
    FlowField,
    FlowSampleUnavailableError,
    fetch_flow_sample,
    search_fields,
    summarize_fields,
)

log = logging.getLogger(__name__)

router = APIRouter(tags=["filter-composer"])

_MAX_FIELDS_RENDUS = 40
"""Au-delà, la liste cesse d'être parcourable à l'œil. La recherche est là pour
réduire — afficher 61 champs d'un coup ne rend service à personne."""

_MAX_VALEURS_PAR_CHAMP = 8


def get_clickhouse_client() -> Any:
    """Placeholder — câblé par `app/main.py`, comme les autres routers."""
    raise RuntimeError(
        "get_clickhouse_client non câblée : app/main.py doit surcharger cette "
        "dépendance avec le client ClickHouse réel."
    )


ClickHouseClient = Annotated[Any, Depends(get_clickhouse_client)]


def _degrade(message: str, detail: str = "") -> HTMLResponse:
    """Rend un état INDÉTERMINÉ, distinct d'un résultat vide.

    « Aucun champ ne correspond à votre recherche » et « je n'ai pas pu lire
    les flux » produiraient sinon le même écran, et l'opérateur conclurait à
    tort que son infrastructure n'émet rien.
    """
    suffixe = f"<p class='composer-degraded__detail'>{escape(detail)}</p>" if detail else ""
    return HTMLResponse(
        f'<div class="notice notice-warn composer-degraded" data-composer-degraded>'
        f"<strong>{escape(message)}</strong>{suffixe}</div>"
    )


def _render_field(field: FlowField) -> str:
    """Une carte de champ, cliquable, avec ses valeurs réellement observées."""
    valeurs = field.sample_values[:_MAX_VALEURS_PAR_CHAMP]
    puces = "".join(
        f'<button type="button" class="composer-value" '
        f'data-composer-field="{escape(field.name)}" '
        f'data-composer-value="{escape(valeur)}">{escape(valeur)}</button>'
        for valeur in valeurs
    )
    if not puces:
        # Champ présent au catalogue mais sans aucune valeur dans l'échantillon :
        # le dire, plutôt que de rendre une carte muette qui laisse croire à un
        # défaut d'affichage.
        puces = (
            '<span class="composer-value composer-value--empty">'
            "aucune valeur dans l'échantillon</span>"
        )

    reste = field.distinct_count - len(valeurs)
    surplus = f'<span class="composer-field__more">+{reste} autre(s)</span>' if reste > 0 else ""

    return (
        f'<li class="composer-field" data-composer-field-card="{escape(field.name)}">'
        f'<button type="button" class="composer-field__head" '
        f'data-composer-field="{escape(field.name)}">'
        f'<span class="composer-field__label">{escape(field.label)}</span>'
        f'<code class="composer-field__name">{escape(field.name)}</code>'
        f"</button>"
        f'<div class="composer-field__values">{puces}{surplus}</div>'
        f"</li>"
    )


@router.get("/config/filters/fields", response_class=HTMLResponse)
def get_filter_fields(client: ClickHouseClient, q: str = "") -> HTMLResponse:
    """Champs de l'échantillon, filtrés par la barre de recherche.

    La recherche porte sur le nom TECHNIQUE, le libellé FRANÇAIS **et les
    valeurs observées** : taper « proxy-frontal » remonte le champ `ExporterName`,
    sans qu'il faille deviner qu'on cherche « exportateur ». C'est ce qui rend
    l'écran utilisable par quelqu'un qui connaît son infrastructure mais pas le
    schéma d'Akvorado.
    """
    try:
        rows = fetch_flow_sample(client)
    except FlowSampleUnavailableError as exc:
        log.error("get_filter_fields: echantillon indisponible: %s", exc)
        return _degrade(
            "Échantillon de flux indisponible.",
            "La lecture des derniers flux n'a pas abouti — la composition par "
            "clic est momentanément hors service. La saisie manuelle reste "
            "possible ci-dessous.",
        )

    if not rows:
        return _degrade(
            "Aucun flux sur la dernière heure.",
            "L'échantillon est vide : il n'y a rien à proposer. Ce n'est pas "
            "une erreur de lecture — la fenêtre observée ne contient aucun flux.",
        )

    fields = summarize_fields(rows)
    if q.strip():
        fields = search_fields(fields, q)

    # Les champs non filtrables n'ont rien à faire dans un compositeur de
    # filtres : les proposer produirait des expressions qu'Akvorado refuse.
    fields = [f for f in fields if f.filterable]

    if not fields:
        return HTMLResponse(
            '<p class="empty" data-composer-no-match>Aucun champ ne correspond à '
            f"« {escape(q)} ». Essayez un nom de machine, un port, un pays…</p>"
        )

    cartes = "".join(_render_field(f) for f in fields[:_MAX_FIELDS_RENDUS])
    reste = len(fields) - _MAX_FIELDS_RENDUS
    note = (
        f'<p class="composer-fields__note">{reste} champ(s) supplémentaire(s) — '
        "affinez la recherche.</p>"
        if reste > 0
        else ""
    )
    return HTMLResponse(f'<ul class="composer-fields">{cartes}</ul>{note}')


@router.post("/config/filters/values", response_class=HTMLResponse)
async def get_filter_values(
    column: Annotated[str, Form()],
    prefix: Annotated[str, Form()] = "",
) -> HTMLResponse:
    """Valeurs proposées pour une colonne, par la console Akvorado.

    Ces valeurs sont lues dans ClickHouse par la console : ce sont donc celles
    RÉELLEMENT observées, pas une liste figée qui dériverait. La console
    indique aussi si la valeur doit être entourée de guillemets — deviner cette
    règle côté client produirait des filtres syntaxiquement faux.
    """
    try:
        completions = await complete_filter("value", column=column, prefix=prefix)
    except ConsoleUnavailable as exc:
        log.error("get_filter_values: console indisponible: colonne=%s: %s", column, exc)
        return _degrade(
            "Suggestions indisponibles.",
            "La console Akvorado n'a pas répondu : les valeurs proposées ne "
            "peuvent pas être listées. La saisie manuelle reste possible.",
        )

    if not completions:
        return HTMLResponse(
            '<p class="empty">Aucune valeur connue pour ce champ dans les flux récents.</p>'
        )

    puces = "".join(
        f'<button type="button" class="composer-value" '
        f'data-composer-field="{escape(column)}" '
        f'data-composer-value="{escape(c.label)}" '
        f'data-composer-quoted="{"1" if c.quoted else "0"}">{escape(c.label)}</button>'
        for c in completions
    )
    return HTMLResponse(f'<div class="composer-field__values">{puces}</div>')


@router.post("/config/filters/validate", response_class=HTMLResponse)
async def post_validate_filter(
    expression: Annotated[str, Form()] = "",
    content: Annotated[str, Form()] = "",
) -> HTMLResponse:
    """Valide une expression AU FIL DE LA FRAPPE, par Akvorado lui-même.

    Le verdict vient de la console, seule autorité sur sa propre grammaire.
    Quand elle refuse, elle indique la POSITION du caractère fautif — c'est ce
    qui distingue un message utile d'un « filtre invalide » qui laisse
    chercher.

    DEUX noms acceptés, et ce n'est pas de la complaisance. La zone de saisie
    est le `<textarea name="content">` du formulaire d'ajout — nom imposé par
    le contrat d'écriture des filtres, qu'on ne peut pas changer. htmx poste
    donc `content`, alors que cette route attendait `expression`.

    DÉFAUT MESURÉ AU NAVIGATEUR (2026-08-06) : le champ arrivait vide, la route
    répondait « filtre vide : aucun filtrage appliqué »… et l'affichait comme
    un SUCCÈS. Une expression fautive aurait donc été annoncée valide, et
    enregistrée telle quelle. Le pire cas d'un validateur : dire oui sans avoir
    rien lu.
    """
    valeur = expression or content
    try:
        result = await validate_filter(valeur)
    except ConsoleUnavailable as exc:
        log.error("post_validate_filter: console indisponible: %s", exc)
        return _degrade(
            "Validation indisponible.",
            "La console Akvorado n'a pas répondu : l'expression n'a PAS été "
            "vérifiée. Elle peut être enregistrée, mais sans garantie de "
            "syntaxe.",
        )

    if result.ok:
        return HTMLResponse(
            '<p class="notice notice-ok composer-verdict" data-composer-valid="1">'
            f"<strong>Expression valide.</strong> {escape(result.message)}</p>"
        )

    return HTMLResponse(
        '<p class="notice notice-crit composer-verdict" data-composer-valid="0">'
        f"<strong>Expression refusée par Akvorado.</strong> {escape(result.message)}</p>"
    )
